variables/variables.go

```diff
@@ -0,0 +1,292 @@
+package variables
+
+import (
+	"fmt"
+	"github.com/gruntwork-io/boilerplate/util"
+)
+
+// An interface for a variable defined in a boilerplate.yml config file
+type Variable interface {
```

Comment:
You make a valid point how a project like this, which now has first-class support for types, seems to fight against the limited type system of Go. I suppose we could consider working with other languages like Scala as well, but then we have all the overhead of setting up our SBT project, figuring out how to do CLI apps in Scala, one more language a future employee has to learn, etc.

So given the options, I think "good enough" is indeed good enough.